File: commons-core/src/main/java/xyz/wagyourtail/commons/core/string/NameType.java
package xyz.wagyourtail.commons.core.string;

import lombok.AllArgsConstructor;
import lombok.val;
import xyz.wagyourtail.commons.core.StringUtils;
import xyz.wagyourtail.commons.core.function.Function;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

@AllArgsConstructor
public class NameType {
    public static NameType CAMEL_CASE = new NameType(
            new Function<String, Iterable<String>>() {
                @Override
                public Iterable<String> apply(final String it) {
                    List<String> list = new ArrayList<>();
                    StringBuilder sb = new StringBuilder();
                    for (int i = 0; i < it.length(); ++i) {
                        if (Character.isUpperCase(it.charAt(i))) {
                            if (sb.length() != 0 && Character.isLowerCase(sb.charAt(sb.length() - 1))) {
                                list.add(sb.toString().toLowerCase(Locale.ROOT));
                                sb.setLength(0);
                            }
                        } else if (sb.length() != 0) {
                            val lastChar = sb.charAt(sb.length() - 1);
                            if (sb.length() > 1 && Character.isUpperCase(lastChar)) {
                                // append multiple capitals and remove them from the sb
                                val s = sb.substring(0, sb.length() - 1);
                                list.add(!s.isEmpty() ? s : s.toLowerCase(Locale.ROOT));
                                sb.delete(0, sb.length() - 1);
                            }
                        }
                        sb.append(it.charAt(i));
                    }
                    if (sb.length() != 0) {
                        if (sb.length() > 1 && Character.isUpperCase(sb.charAt(0)) && Character.isUpperCase(sb.charAt(1))) {
                            list.add(sb.toString());
                        } else {
                            list.add(sb.toString().toLowerCase(Locale.ROOT));
                        }
                    }
                    return list;
                }
            },
            new Function<Iterable<String>, String>() {
                @Override
                public String apply(Iterable<String> strings) {
                    val iter = strings.iterator();
                    StringBuilder sb = new StringBuilder(iter.next());
                    while (iter.hasNext()) {
                        sb.append(StringUtils.capitalize(iter.next()));
                    }
                    return sb.toString();
                }
            }
    );

    public static NameType PASCAL_CASE = new NameType(
            CAMEL_CASE.split,
            new Function<Iterable<String>, String>() {
                @Override
                public String apply(Iterable<String> strings) {
                    val iter = strings.iterator();
                    StringBuilder sb = new StringBuilder();
                    while (iter.hasNext()) {
                        sb.append(StringUtils.capitalize(iter.next()));
                    }
                    return sb.toString();
                }
            }
    );

    public static NameType SNAKE_CASE = new NameType(
            new Function<String, Iterable<String>>() {
                @Override
                public Iterable<String> apply(String s) {
                    return Arrays.asList(s.split("_"));
                }
            },
            new Function<Iterable<String>, String>() {
                @Override
                public String apply(Iterable<String> strings) {
                    return StringUtils.joinToString("_", strings);
                }
            }
    );

    public static NameType KEBAB_CASE = new NameType(
            new Function<String, Iterable<String>>() {
                @Override
                public Iterable<String> apply(String s) {
                    return Arrays.asList(s.split("-"));
                }
            },
            new Function<Iterable<String>, String>() {
                @Override
                public String apply(Iterable<String> strings) {
                    return StringUtils.joinToString("-", strings);
                }
            }
    );

    private final Function<String, Iterable<String>> split;
    private final Function<Iterable<String>, String> join;


    public static String convert(String value, NameType from, NameType to) {
        return from.convert(to, value);
    }

    public String convert(NameType to, String value) {
        if (to == this) {
            return value;
        }

        if (value.isEmpty()) {
            return value;
        }

        // shortcuts
        // - pascal/camel
        if (this == PASCAL_CASE && to == CAMEL_CASE) {
            if (
                    (value.length() >= 3 && Character.isUpperCase(value.charAt(0)) && Character.isUpperCase(value.charAt(1)) && Character.isUpperCase(value.charAt(2))) ||
                            (value.length() == 2 && Character.isUpperCase(value.charAt(0)) && Character.isUpperCase(value.charAt(1)))
            ) {
                return value;
            } else {
                return StringUtils.uncapitalize(value);
            }
        }
        if (this == CAMEL_CASE && to == PASCAL_CASE) {
            return StringUtils.capitalize(value);
        }

        // - kebab/snake
        if (this == KEBAB_CASE && to == SNAKE_CASE) {
            return value.replace("-", "_");
        }
        if (this == SNAKE_CASE && to == KEBAB_CASE) {
            return value.replace("_", "-");
        }

        // fallback
        Iterable<String> split = this.split.apply(value);
        return to.join.apply(split);
    }

}

File: commons-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/string/NameType.kt
package xyz.wagyourtail.commonskt.string

import xyz.wagyourtail.commonskt.utils.capitalized
import xyz.wagyourtail.commonskt.utils.iterable
import xyz.wagyourtail.commonskt.utils.uncapitalized

fun String.convertNameType(from: NameType, to: NameType): String {
    if (from == to) {
        return this
    }

    // shortcuts

    // - pascal/camel
    if (from == NameType.PASCAL_CASE && to == NameType.CAMEL_CASE) {
        // check if starts with initialism
        return if (
            (this.length >= 3 && this[0].isUpperCase() && this[1].isUpperCase() && this[2].isUpperCase()) ||
            (this.length == 2 && this[0].isUpperCase() && this[1].isUpperCase())
        ) {
            this
        } else {
            this.uncapitalized()
        }
    }
    if (from == NameType.CAMEL_CASE && to == NameType.PASCAL_CASE) {
        return this.capitalized()
    }

    // - kebab/snake
    if (from == NameType.KEBAB_CASE && to == NameType.SNAKE_CASE) {
        return this.replace('-', '_')
    }
    if (from == NameType.SNAKE_CASE && to == NameType.KEBAB_CASE) {
        return this.replace('_', '-')
    }

    // fallback
    val split = from.split(this)
    return to.join(split)
}

/**
 * @param split should return an iterable of the lowercase words making up the name (with initalisms remaining capitalized)
 * @param join should join the iterable of words into a string
 */
data class NameType(val split: (String) -> Iterable<String>, val join: (Iterable<String>) -> String) {

    companion object {
        val CAMEL_CASE = NameType(
            {
                // split by upper case, keep multiple capitals in a row together
                iterable {
                    val sb = StringBuilder()
                    for (i in 0 until it.length) {
                        if (it[i].isUpperCase()) {
                            if (sb.isNotEmpty() && sb[sb.length - 1].isLowerCase()) {
                                yield(sb.toString().lowercase())
                                sb.clear()
                            }
                            sb.append(it[i])
                        } else {
                            if (sb.isNotEmpty()) {
                                val lastChar = sb[sb.length - 1]
                                if (sb.length > 1 && lastChar.isUpperCase()) {
                                    // append multiple capitals and remove them from the sb
                                    val s = sb.substring(0, sb.length - 1)
                                    yield(if (s.length > 1) s else s.lowercase())
                                    sb.deleteRange(0, sb.length - 1)
                                }
                            }
                            sb.append(it[i])
                        }
                    }
                    if (sb.isNotEmpty()) {
                        if (sb.length > 1 && sb[0].isUpperCase() && sb[1].isUpperCase()) {
                            yield(sb.toString())
                        } else {
                            yield(sb.toString().lowercase())
                        }
                    }
                }
            },
            {
                val iter = it.iterator()
                iter.next() + iter.asSequence().joinToString("") { it.capitalized() }
            }
        )

        val PASCAL_CASE = NameType(
            CAMEL_CASE.split,
            { CAMEL_CASE.join(it).capitalized() }
        )

        val SNAKE_CASE = NameType(
            { it.split("_") },
            { it.joinToString("_") }
        )

        val KEBAB_CASE = NameType(
            { it.split("-") },
            { it.joinToString("-") }
        )

    }

}